prover/crates/bin/witness_generator/src/basic_circuits.rs

```diff
@@ -446,43 +461,103 @@ async fn generate_witness(
                 let parent_span = span.clone();
                 tracing::info_span!(parent: parent_span, "send_circuit").in_scope(|| {
-                    circuit_sender.blocking_send(circuit).unwrap();
+                    circuit_sender
+                        .blocking_send(circuit)
+                        .expect("failed to send circuit from harness");
                 });
             },
-            |a, b, c| queue_sender.blocking_send((a as u8, b, c)).unwrap(),
+            |a, b, c| {
+                queue_sender
+                    .blocking_send((a as u8, b, c))
+                    .expect("failed to send recursion queue from harness")
+            },
         );
         (scheduler_witness, block_witness)
     })
     .instrument(make_circuits_span);
 
-    let mut circuit_urls = vec![];
-    let mut recursion_urls = vec![];
+    let mut save_circuit_handles = vec![];
 
-    let mut circuits_present = HashSet::<u8>::new();
+    // Ordering determines how we compose the circuit proofs in Leaf Aggregation Round.
+    // Sequence is used to determine circuit ordering (the sequencing of instructions) .
+    // If the order is tampered with, proving will fail (as the proof would be computed for a different sequence of instruction).
+    let circuit_sequence = AtomicUsize::new(0);
 
     let save_circuits_span = tracing::info_span!("save_circuits");
-    let save_circuits = async {
-        loop {
-            tokio::select! {
-                Some(circuit) = circuit_receiver.recv().instrument(tracing::info_span!("wait_for_circuit")) => {
-                    circuits_present.insert(circuit.numeric_circuit_type());
-                    circuit_urls.push(
-                        save_circuit(block_number, circuit, circuit_urls.len(), object_store).await,
-                    );
-                }
-                Some((circuit_id, queue, inputs)) = queue_receiver.recv().instrument(tracing::info_span!("wait_for_queue")) => {
-                    let urls = save_recursion_queue(block_number, circuit_id, queue, &inputs, object_store).await;
-                    recursion_urls.push(urls);
-                }
-                else => break,
-            };
+
+    // Future which receives circuits and saves them async.
+    let circuit_receiver_handle = async {
+        let semaphore = Arc::new(Semaphore::new(MAX_IN_FLIGHT_CIRCUITS));
+
+        while let Some(circuit) = circuit_receiver
+            .recv()
+            .instrument(tracing::info_span!("wait_for_circuit"))
+            .await
+        {
+            let sequence = circuit_sequence.fetch_add(1, Ordering::SeqCst);
+            let object_store = object_store.clone();
+            let semaphore = semaphore.clone();
+            save_circuit_handles.push(tokio::task::spawn(async move {
+                let _permit = semaphore
+                    .acquire()
+                    .await
+                    .expect("failed to get permit for running save circuit task");
```

Comment:
I don't think this works. You're limiting the number of spawned jobs, but not the number of created futures.
Imagine that we have only 1 permit.
We receive a new circuit, we spawn a task. When it's launched, we get a permit. 1 job in progress, 1 object in memory.
We receive the next circuit, we spawn a task. When it's launched, we can't get a permit. 1 job in progress, 2 objects in memory.
We receiver the third circuit, we spawn a task. When it's launched, we can't get a permit. 1 job in progress, 3 objects in memory. Etc.

For it to work, you need to obtain an [owned permit](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#method.acquire_owned) before you spawn the task, move it into the task, and drop it after `save_circuits`. This way lack of permits would stop us from letting witness generation to proceed.